File: server/actions/users/usersProfile.py
# -*- coding: utf-8 -*-
import inject
import logging
import uuid
# import asyncio
# from asyncio import coroutine
# from autobahn.asyncio.wamp import ApplicationSession

#from model.users.usersProfileModel import UsersProfileModel
from model.users.entities.user import User
from model.users.entities.mail import Mail
from model.users.usersModel import UsersModel



# from model.exceptions import *

import autobahn
import wamp

class UsersProfile(wamp.SystemComponentSession):

    def getRegisterOptions(self):
        return autobahn.wamp.RegisterOptions(details_arg='details')

    @autobahn.wamp.register('users.profile.find_by_id')
    def findById(self, id, details):
        #administracion de usuario
        ctx = wamp.getContextManager()
        ctx.getConn()
        try:
            return User.findById(ctx, id)

        finally:
            ctx.closeConn()

    @autobahn.wamp.register('users.profile.find_emails_by_user_id')
    def findEmailsByUserId(self, userId, details):
        ctx = wamp.getContextManager()
        ctx.getConn()
        try:
            return Mail.find(ctx, userId=userId).fetch(ctx)

        finally:
            ctx.closeConn()


    @autobahn.wamp.register('users.profile.add_email')
    def addEmail(self, email, details):
        ctx = wamp.getContextManager()
        ctx.getConn()
        try:
            email.persist(ctx)
            ctx.con.commit()
            return email

        finally:
            ctx.closeConn()

    @autobahn.wamp.register('users.profile.delete_email')
    def deleteEmail(self, email, details):
        ctx = wamp.getContextManager()
        ctx.getConn()
        try:
            email.delete(ctx)
            ctx.con.commit()

        finally:
            ctx.closeConn()

    @autobahn.wamp.register('users.profile.send_email_confirmation')
    def sendEmailConfirmation(self, userId, eid, details):
        ctx = wamp.getContextManager()
        ctx.getConn()
        try:
            logging.warn('userId: {} sendEmailConfirmation {}'.format(userId,eid))
            UsersModel.sendEmailConfirmation(ctx, userId, eid)
            ctx.con.commit()

        finally:
            ctx.closeConn()

    @autobahn.wamp.register('users.profile.process_code')
    def processCode(self, emailId, code, details):
        ctx = wamp.getContextManager()
        ctx.getConn()
        try:
            emails = Mail.findByIds(ctx, [emailId])
            if len(emails) <= 0:
                raise Exception('No existe el correo')
            if code != emails[0].hash:
                raise Exception('Código incorrecto')

            emails[0].confirmed = True
            emails[0].persist(ctx)
            ctx.con.commit()
        finally:
            ctx.closeConn()

    @autobahn.wamp.register('users.profile.persist')
    def persist(self, user, details):
        #administracion de usuario
        ctx = wamp.getContextManager()
        ctx.getConn()
        try:
            user.persist(ctx)
            ctx.con.commit()

        finally:
            ctx.closeConn()



    @autobahn.wamp.register('users.profile.change_password')
    def changePassword(self, password, details):
        ctx = wamp.getContextManager()
        ctx.getConn()
        try:
            users = self.getUserId(ctx, details)
            logging.info(password)
            userId = users[0].id if len(users) > 0 else None
            logging.info(userId)
            r = UsersModel.changePassword(ctx, userId, password)
            ctx.con.commit()
            return r


        finally:
            ctx.closeConn()

File: web/systems/users/profile/services/UsersProfile.js
(function() {
    'use strict'
    angular
      .module('users.profile')
      .service('UsersProfile',UsersProfile);

    UsersProfile.inject = ['$rootScope', 'Login', '$q', 'Files'];

    function UsersProfile($rootScope, Login, $q, Files) {

      //Buscar usuario
      //@param id de usuario o null (si esta definido el id buscara el usuario en la base)
      this.findById = function(id){ return Login.getPrivateTransport().call('users.profile.find_by_id', [id]); }

      //Persistir usuario
      //@param user User a persistir
      this.persist = function(user){ return Login.getPrivateTransport().call('users.profile.persist', [user]); }

      //Buscar emails de usuario
      //@param user User a persistir
      this.findEmailsByUserId = function(userId){ return Login.getPrivateTransport().call('users.profile.find_emails_by_user_id', [userId]); }

      //Agregar email
      //@param email Email a persistir
      this.addEmail = function(email){ return Login.getPrivateTransport().call('users.profile.add_email', [email]); }

      //Eliminar email
      //@param email Email a eliminar
      this.deleteEmail = function(email){ return Login.getPrivateTransport().call('users.profile.delete_email', [email]); }

      //Enviar codigo de confirmacion por email
      this.sendCode = function(userId){ return Login.getPrivateTransport().call('users.profile.send_code', [userId]); }

      //Procesar codigo de confirmacion
      this.processCode = function(code){ return Login.getPrivateTransport().call('users.profile.process_code', [code]); }

  }
})();
